add functions to create carbon file from array_it

Jasper-Ben opened the pull request:
Adds functions for creating a carbon file from an array iterator.
Cherry-picked commit https://github.com/jaksonlabs/jakson/pull/4/commits/442493207ded119bef22e21bb1e850c9e5b640b7, so you might want to review https://github.com/jaksonlabs/jakson/pull/4 first.
Also cherry-picked commit 656b21689f7295f3726bb9f81571f8a300ca1661 from https://github.com/jaksonlabs/jakson/pull/1.

Jasper-Ben committed:
added function carbon_object_it_length

returns the number of objects in object iterator

Jasper-Ben committed:
add functions to check for atomar field types

these functions return false when the field type is a column containing items of the desired type

Jasper-Ben committed:
add functions to create carbon file from array it

currently all signed and unsigned values are saved as 64 bit values.
This should be extended to handle other sizes as well.

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+    while (carbon_object_it_next(it_in)) {
+        u64 keylen;
+        const char *_key = carbon_object_it_prop_name(&keylen, it_in);
+        const char *key = strndup(_key, keylen);
```
memleak, fixme

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+        else if (carbon_field_type_is_string(type)) {
+            u64 strlen;
+            const char *_str = carbon_object_it_string_value(&strlen, it_in);
+            const char *str = strndup(_str, strlen);
```
memleak, fixme

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+        else if (carbon_field_type_is_binary(type)) {
+            carbon_binary bin;
+            carbon_object_it_binary_value(&bin, it_in);
+            const char *mime = strndup(bin.mime_type, bin.mime_type_strlen);
```
memleak, fixme

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+        else if (carbon_field_type_is_string(type)) {
+            u64 strlen;
+            const char *_str = carbon_array_it_string_value(&strlen, it_in);
+            const char *str = strndup(_str, strlen);
```
memleak, fixme

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+        else if (carbon_field_type_is_binary(type)) {
+            carbon_binary bin;
+            carbon_array_it_binary_value(&bin, it_in);
+            const char *mime = strndup(bin.mime_type, bin.mime_type_strlen);
```
memleak, fixme

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+                        carbon_insert_null(nested_ins);
+                    }
+                    else {
+                        carbon_insert_i64(nested_ins, *((i64*)&(vals[i])));
```
this will break, as it will try to open i8 mem as i64 etc.

heregoesmarcel commented:
Currently, your added `carbon_from_array_it()` seems to initialize a new carbon document with a fixed key and options. This should to my knowledge be parameterized.

Jasper-Ben commented on src/jakson/carbon.c:
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
+            carbon_insert *nested_ins;
+            carbon_insert_array_state state;
+
+            nested_ins = carbon_insert_array_begin(&state, ins, 1024);
```
use array_len param, instead of fixed size

Jasper-Ben committed:
improvements to insert from iter. functions

- fixed memory leaks caused by strndup
- changed reserved memory size for nested data according to nested data
size, instead of fixed value
